File: 7kyu/chain_me.py
# CHAIN ME
# Write a generic function chainer that takes a starting value, and an array of functions to execute on it (array of symbols for Ruby).

# The input for each function is the output of the previous function (except the first function, which takes the starting value as its input). Return the final value after execution is complete.

# def add10(x): return x + 10
# def mul30(x): return x * 30

# chain(50, [add10, mul30])
# returns 1800

# MY SOLUTION
def chain(init_val, functions):
    val = init_val
    for function in functions:
        val = function(val)
    return val

# OTHER CLEVER SOLUTION
from functools import reduce

def chain(init_val, functions):
    return reduce(lambda x, f: f(x), functions, init_val)

File: 7kyu/chainMe.js
//CHAIN ME
// Write a generic function chainer that takes a starting value, and an array of functions to execute on it (array of symbols for Ruby).

// The input for each function is the output of the previous function (except the first function, which takes the starting value as its input). Return the final value after execution is complete.

// function add(num) {
//   return num + 1;
// }

// function mult(num) {
//   return num * 30;
// }

// chain(2, [add, mult]);
// returns 90;

//MY SOLUTION
function chain(num, funcArr){
    let feed = [num];
    for(i = 0; i < funcArr.length; i++){
        feed = feed.map(x => funcArr[i](x));
    }
    return feed[0];
}

//OTHER CLEVER SOLUTION
function chain(v, fns) {
    return fns.reduce(function(v, fn) { return fn(v) }, v);
}